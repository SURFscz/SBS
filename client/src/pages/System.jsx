import React from "react";
import "./System.scss";
import I18n from "i18n-js";
import {auditLogsActivity, clearAuditLogs, dbSeed, dbStats, health, outstandingRequests, suspendUsers} from "../api";
import Button from "../components/Button";
import {isEmpty} from "../utils/Utils";
import UnitHeader from "../components/redesign/UnitHeader";
import ConfirmationDialog from "../components/ConfirmationDialog";
import SpinnerField from "../components/redesign/SpinnerField";
import {AppStore} from "../stores/AppStore";
import Tabs from "../components/Tabs";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import Activity from "../components/Activity";
import Select from "react-select";
import MemberJoinRequests from "../components/redesign/MemberJoinRequests";
import MemberCollaborationRequests from "../components/redesign/MemberCollaborationRequests";

const options = [25, 50, 100].map(nbr => ({value: nbr, label: nbr}));

class System extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            tab: "cron",
            suspendedUsers: {},
            outstandingRequests: {},
            databaseStats: [],
            seedResult: null,
            confirmationDialogOpen: false,
            confirmationDialogQuestion: undefined,
            confirmationDialogAction: () => true,
            cancelDialogAction: () => this.setState({confirmationDialogOpen: false}),
            busy: false,
            auditLogs: {audit_logs: []},
            filteredAuditLogs: {audit_logs: []},
            limit: options[1],
            query: ""
        }
    }

    componentDidMount = () => {
        health().then(() => {
            const params = this.props.match.params;
            const tab = params.tab || this.state.tab;
            this.tabChanged(tab);
            AppStore.update(s => {
                s.breadcrumb.paths = [
                    {path: "/", value: I18n.t("breadcrumb.home")},
                    {value: I18n.t("breadcrumb.system")}
                ];
            });

        })
    };

    clear = () => {
        this.setState({
            suspendedUsers: {},
            outstandingRequests: {},
            databaseStats: [],
            seedResult: null,
            query: "",
            auditLogs: {audit_logs: []},
            filteredAuditLogs: {audit_logs: []},
            limit: options[1],
        });
    }

    reload = () => {
        // eslint-disable-next-line
        window.location.href = window.location.href;
    }

    getCronTab = (suspendedUsers, outstandingRequests) => {
        return (<div key="cron" name="cron" label={I18n.t("home.tabs.cron")}
                     icon={<FontAwesomeIcon icon="clock"/>}>
            <div className="mod-system">
                <section className={"info-block-container"}>
                    {this.renderDailyCron()}
                    {this.renderDailyCronResults(suspendedUsers)}
                    {this.renderOutstandingRequests()}
                    {this.renderOutstandingRequestsResults(outstandingRequests)}
                </section>
            </div>
        </div>)
    }

    changeLimit = val => {
        this.setState({limit: val, busy: true}, () => {
            auditLogsActivity(val.value).then(res => {
                this.setState({
                    auditLogs: res,
                    filteredAuditLogs: this.filterAuditLogs(res, this.state.query),
                    busy: false
                });
            });
        });
    }

    filterAuditLogs = (auditLogs, query) => {
        if (isEmpty(query)) {
            return auditLogs;
        }
        const lowerQuery = query.toLowerCase();
        const sub = [...auditLogs.audit_logs].filter(a => {
            let matchesParent = false;
            let matchesUser = false;
            let matchesName = false
            if (a.parent_name && auditLogs[a.parent_name]) {
                const parent = auditLogs[a.parent_name].find(obj => obj.id === a.parent_id);
                if (parent && parent.name) {
                    matchesParent = parent.name.toLowerCase().indexOf(lowerQuery) > -1;
                }
            }
            if (a.subject_id && auditLogs.users) {
                const subject = auditLogs.users.find(user => user.id === a.subject_id);
                if (subject && subject.name) {
                    matchesUser = subject.name.toLowerCase().indexOf(lowerQuery) > -1;
                }
            }
            if (a.target_name) {
                matchesName = a.target_name.toLowerCase().indexOf(lowerQuery) > -1;
            }
            return matchesParent || matchesUser || matchesName;
        });
        return {...auditLogs, audit_logs: sub};
    }

    getActivityTab = (filteredAuditLogs, limit, query, config) => {
        return (
            <div key="activity" name="activity" label={I18n.t("home.tabs.activity")}
                 icon={<FontAwesomeIcon icon="code-branch"/>}>
                <div className="mod-system">
                    <section className={"info-block-container"}>
                        <section className="search-activity">
                            <p>{I18n.t("system.activity")}</p>
                            {config.seed_allowed &&
                            <Button warningButton={true} onClick={() => this.doClearAuditLogs(true)}/>}
                            <div className={`search ${config.seed_allowed ? "" : "no-clear-logs"}`}>
                                <input type="text"
                                       onChange={this.onChangeQuery}
                                       value={query}
                                       placeholder={I18n.t("system.searchPlaceholder")}/>
                                <FontAwesomeIcon icon="search"/>
                            </div>
                            <Select
                                className="input-select-inner"
                                classNamePrefix="select-inner"
                                value={limit}
                                placeholder={I18n.t("system.searchPlaceholder")}
                                onChange={this.changeLimit}
                                options={options}
                                isSearchable={false}
                                isClearable={false}
                            />
                        </section>
                        <Activity auditLogs={filteredAuditLogs}/>
                    </section>
                </div>
            </div>)
    }

    onChangeQuery = e => {
        const query = e.target.value;
        const {auditLogs} = this.state;
        const filteredAuditLogs = this.filterAuditLogs(auditLogs, query);
        this.setState({
            filteredAuditLogs: filteredAuditLogs,
            query: query
        });
    }

    getDatabaseTab = databaseStats => {
        return (<div key="database" name="database" label={I18n.t("home.tabs.database")}
                     icon={<FontAwesomeIcon icon="database"/>}>
            <div className="mod-system">
                <section className={"info-block-container"}>
                    {this.renderDbStats()}
                    {this.renderDbStatsResults(databaseStats)}
                </section>
            </div>
        </div>)
    }

    getSeedTab = seedResult => {
        return (<div key="seed" name="seed" label={I18n.t("home.tabs.seed")}
                     icon={<FontAwesomeIcon icon="seedling"/>}>
            <div className="mod-system">
                <section className={"info-block-container"}>
                    {this.renderDbSeed()}
                    <p className="result">{seedResult}</p>
                </section>
            </div>
        </div>)
    }

    confirm = (action, question) => {
        this.setState({
            confirmationDialogOpen: true,
            confirmationDialogQuestion: question,
            confirmationDialogAction: action
        });
    };

    doSuspendUsers = () => {
        this.setState({busy: true})
        suspendUsers().then(res => {
            this.setState({suspendedUsers: res, busy: false});
        });
    }

    doOutstandingRequests = () => {
        this.setState({busy: true})
        outstandingRequests().then(res => {
            // debugger;
            this.setState({outstandingRequests: res, busy: false});
        });
    }

    doClearAuditLogs = showConfirmation => {
        if (showConfirmation) {
            this.confirm(() => this.doClearAuditLogs(false), I18n.t("system.runClearAuditLogsConfirmation"));
        } else {
            this.setState({confirmationDialogOpen: false, busy: true,});
            clearAuditLogs().then(() => this.setState({
                auditLogs: {audit_logs: []},
                filteredAuditLogs: {audit_logs: []},
                busy: false
            }));
        }

    }

    doDbSeed = showConfirmation => {
        if (showConfirmation) {
            this.confirm(() => this.doDbSeed(false), I18n.t("system.runDbSeedConfirmation"));
        } else {
            this.setState({confirmationDialogOpen: false, busy: true,});
            const d = new Date();
            dbSeed().then(() => {
                this.setState({
                    busy: false,
                    seedResult: I18n.t("system.seedResult", {ms: new Date().getMilliseconds() - d.getMilliseconds()})
                });
            });
        }
    }

    renderDailyCron = () => {
        const {suspendedUsers} = this.state;
        return (
            <div className="info-block">
                <p>{I18n.t("system.runDailyJobsInfo")}</p>
                <div className="actions">
                    {isEmpty(suspendedUsers) && <Button txt={I18n.t("system.runDailyJobs")}
                                                        onClick={this.doSuspendUsers}/>}
                    {!isEmpty(suspendedUsers) && <Button txt={I18n.t("system.clear")}
                                                         onClick={this.clear} cancelButton={true}/>}
                </div>
            </div>
        );
    }

    renderOutstandingRequests = () => {
        const {outstandingRequests} = this.state;
        return (
            <div className="info-block">
                <p>{I18n.t("system.runOutdatedRequestsInfo")}</p>
                <div className="actions">
                    {isEmpty(outstandingRequests) && <Button txt={I18n.t("system.runOutdatedRequests")}
                                                             onClick={this.doOutstandingRequests}/>}
                    {!isEmpty(outstandingRequests) && <Button txt={I18n.t("system.clear")}
                                                              onClick={this.clear} cancelButton={true}/>}
                </div>
            </div>
        );
    }

    renderDbStats = () => {
        return (
            <div className="info-block">
                <p>{I18n.t("system.runDbStatsInfo")}</p>
            </div>
        );
    }

    renderDbSeed = () => {
        const {seedResult} = this.state;
        return (
            <div className="info-block">
                <p>{I18n.t("system.runDbSeedInfo")}</p>
                <div className="actions">
                    {isEmpty(seedResult) && <Button txt={I18n.t("system.runDbSeed")}
                                                    onClick={() => this.doDbSeed(true)}/>}
                    {!isEmpty(seedResult) && <Button txt={I18n.t("system.reload")}
                                                     onClick={this.reload} cancelButton={true}/>}
                </div>
            </div>
        );
    }

    renderDailyCronResults = suspendedUsers => {
        return (
            <div className="results">
                {!isEmpty(suspendedUsers) && <div className="results">
                    <table className="suspended-users">
                        <thead>
                        <th>{I18n.t("system.action")}</th>
                        <th>{I18n.t("system.results")}</th>
                        </thead>
                        <tbody>
                        {Object.keys(suspendedUsers).map(key =>
                            <tr key={key}>
                                <td className="action">{I18n.t(`system.${key}`)}</td>
                                <td>
                                    {!isEmpty(suspendedUsers[key]) && <ul>
                                        {suspendedUsers[key].map(email => <li>{email}</li>)}
                                    </ul>}
                                    {isEmpty(suspendedUsers[key]) && <span>None</span>}
                                </td>
                            </tr>
                        )}
                        </tbody>
                    </table>
                </div>}
            </div>)
    }

    renderOutstandingRequestsResults = outstandingRequests => {
        const collaboration_requests = outstandingRequests.collaboration_requests;
        const join_requests = outstandingRequests.collaboration_join_requests;
        return (
            <div className="results">
                {!isEmpty(outstandingRequests) && <div className="results">
                    <MemberJoinRequests join_requests={join_requests} isPersonal={false} {...this.props} />
                    <MemberCollaborationRequests {...this.props} isPersonal={false} collaboration_requests={collaboration_requests}/>
                </div>}
            </div>
        )
    }

    renderDbStatsResults = databaseStats => {
        return (
            <div className="results">
                {!isEmpty(databaseStats) && <div className="results">
                    <table className="table-counts">
                        <thead>
                        <tr>
                            <th>{I18n.t("system.name")}</th>
                            <th>{I18n.t("system.count")}</th>
                        </tr>
                        </thead>
                        <tbody>
                        {databaseStats.map((stat, i) => <tr key={i}>
                            <td>{stat.name}</td>
                            <td>{stat.count}</td>
                        </tr>)}
                        </tbody>
                    </table>
                </div>}</div>)
    }

    tabChanged = name => {
        this.clear();
        this.setState({tab: name, busy: true}, () => {
            this.props.history.replace(`/system/${name}`);
        });
        if (name === "database") {
            dbStats().then(res => {
                this.setState({databaseStats: res, busy: false});
            });
        } else if (name === "activity") {
            const {limit} = this.state;
            auditLogsActivity(limit.value).then(res => {
                this.setState({
                    auditLogs: res,
                    filteredAuditLogs: this.filterAuditLogs(res, this.state.query),
                    busy: false
                });
            });
        } else {
            this.setState({busy: false});
        }
    }

    render() {
        const {admin} = this.props.user;
        if (!admin) {
            return null;
        }
        const {
            seedResult, confirmationDialogOpen, cancelDialogAction, confirmationDialogAction, outstandingRequests,
            confirmationDialogQuestion, busy, tab, filteredAuditLogs, databaseStats, suspendedUsers,
            limit, query
        } = this.state;
        const {config} = this.props;

        if (busy) {
            return <SpinnerField/>
        }
        const tabs = [
            this.getCronTab(suspendedUsers, outstandingRequests),
            config.seed_allowed ? this.getSeedTab(seedResult) : null,
            this.getDatabaseTab(databaseStats),
            this.getActivityTab(filteredAuditLogs, limit, query, config)
        ]
        return (
            <div className="mod-system-container">
                <ConfirmationDialog isOpen={confirmationDialogOpen}
                                    cancel={cancelDialogAction}
                                    confirm={confirmationDialogAction}
                                    isWarning={true}
                                    question={confirmationDialogQuestion}/>
                <UnitHeader obj={({name: I18n.t("system.title"), icon: "toolbox"})}/>

                <Tabs activeTab={tab} tabChanged={this.tabChanged}>
                    {tabs}
                </Tabs>


            </div>);
    }
}

export default System;