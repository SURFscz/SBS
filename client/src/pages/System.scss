@import "../stylesheets/vars.scss";

.mod-system-container {
    width: 100%;

    .mod-system {
        width: 100%;
        margin: 0 auto;
        max-width: $medium;
        padding: 0;

        @media (max-width: $compact-medium) {
            padding: 0 15px;
        }


        .info-block-container {
            margin: 15px auto 15px auto;
            display: flex;
            align-items: stretch;
            flex-direction: column;

            p {
                margin: 5px 0 20px 0;

                &.title {
                    font-weight: bolder;
                }
            }

            @media (max-width: $tablet-max) {
                flex-direction: column;
            }

            .info-block {
                border-bottom: solid 4px $lightest-grey;
                padding: 15px 0;
                width: 100%;

                display: flex;
                flex-direction: row;
                align-items: center;

            }

            .actions {
                margin-left: auto;

                button:not(:first-child) {
                    margin-left: 25px;
                }

            }
        }

        p.result {
            margin: 10px 0;
            color: $dark-green;
        }

        .results {
            margin-top: 15px;

            p.sub-title {
                font-size: 20px;
                margin-bottom: 15px;

            }

            ul {
                list-style: initial;

                li {
                    margin: 10px 0 5px 20px;

                    &:first-child {
                        margin-top: 0;
                    }
                }
            }

            span {
                display: inline-block;
                margin: 10px 0;

                &.sds--checkbox--visual {
                    display: flex;
                    margin: 0;
                }
            }

            table {
                width: 100%;
                margin-top: 25px;

                thead {
                    tr {
                        border-bottom: 3px solid $lighter-grey;
                    }

                    th {
                        padding: 10px 0;
                        color: $primary-orange;
                        text-align: left;
                        width: 50%;

                        &.invitation-reminders {
                            width: 33%;
                        }
                    }
                }

                tbody {
                    tr {
                        color: black;
                        border-bottom: 1px solid $lighter-grey;

                        &:nth-child(odd) {
                            background-color: $lightest-grey;
                        }

                        td {
                            padding: 10px 0;

                            &.action {
                                vertical-align: top;
                            }
                        }
                    }
                }
            }

            table.table-logins {

                th {
                    &.type {
                        width: 50%;
                    }

                    &.total {
                        width: 15%;
                        text-align: center;
                    }

                    &.succeeded {
                        width: 15%;
                        text-align: center;
                    }

                    &.failed {
                        width: 15%;
                        text-align: center;
                    }

                }

                tbody tr td {
                    &.count, &.succeeded, &.failed {
                        text-align: center;
                    }

                }

            }
        }

        .search-activity {
            display: flex;
            align-items: center;
            margin-top: 10px;
            justify-content: space-between;

            @media (max-width: $compact-medium) {
                flex-direction: column;
                align-items: normal;
                padding: 0 15px;
            }

            p {
                font-weight: bold;
                @media (max-width: $compact-medium) {
                    margin-bottom: 15px
                }

            }

            button.sds--delete {
                margin: 0 25px 0 auto;

                @media (max-width: $compact-medium) {
                    margin: 0 0 15px 0;
                }
            }

            div.input-select-inner {
                margin-left: 25px;
                min-width: 100px;
                @media (max-width: $compact-medium) {
                    margin-top: 15px;
                    margin-left: 0;
                }
            }
        }

        .delete-all {
            display: flex;
            margin: 25px;

        }
    }

    .table-selection {
        display: flex;

        .select-field {
            width: 100%;
            margin-top: 0;
        }

        div.action-container {
            display: flex;

            button {
                margin: auto 0 0 15px;
            }
        }

    }

    .toggle-json {
        display: flex;
        margin: 15px 0;

        .copy-to-clipboard {
            margin-left: 25px;
        }
    }

    .search {
        position: relative;
        display: flex;

        &.no-clear-logs {
            margin-left: auto;
        }

        @media (max-width: $compact-medium) {
            margin-left: 0;
        }

        svg.fa-search, svg.fa-magnifying-glass {
            position: absolute;
            font-size: 16px;
            color: $blue;
            top: 7px;
            right: 9px;
        }

        &.server-side {
            height: 48px;
            margin-top: auto;
            margin-left: 15px;

            svg.fa-search, svg.fa-magnifying-glass {
                top: 14px;
            }
        }

        input[type=text] {
            flex-grow: 2;
            border: 1px solid $lighter-grey;
            padding: 0 15px 0 10px;
            min-height: 38px;
            font-size: 16px;
            border-radius: $br;
            min-width: 560px;

            @media (max-width: $compact-medium) {
                min-width: 0;
            }


            &:focus {
                outline: none;
                box-shadow: 1px 1px 1px $blue-hover, -1px -1px 1px $blue-hover;
            }
        }

    }

    table.suspended-users {

        margin-bottom: 25px;

        thead {
            th {
                &.name {
                    width: 25%;
                }

                &.email {
                    width: 25%;
                }

                &.lastLogin {
                    width: 25%;
                }

                &.actions {
                    width: 25%;
                }
            }
        }

        tbody {
            td {
                vertical-align: middle;
            }
        }
    }

    .invitation_reminders {
        display: flex;
        flex-direction: column;
    }

}
