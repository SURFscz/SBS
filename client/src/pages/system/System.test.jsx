import React from "react";
import {render} from "@testing-library/react";
import System from "./System";

jest.mock("@fortawesome/react-fontawesome", () => ({
    FontAwesomeIcon: () => <span data-testid="icon"/>
}));

const mockUser = {admin: true};
const mockConfig = {seed_allowed: true};

describe("System", () => {
    it("matches snapshot with default props", () => {
        const {asFragment} = render(
            <System
                user={mockUser}
                config={mockConfig}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
