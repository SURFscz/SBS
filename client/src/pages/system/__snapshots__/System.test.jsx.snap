// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`System matches snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="mod-system-container"
  >
    <div
      class="unit-header-container"
    >
      <div
        class="unit-header"
      >
        <div
          class="image "
        >
          <span
            data-testid="icon"
          />
        </div>
        <div
          class="obj-name"
        >
          <div
            class="meta-info-container"
          >
            <div
              class="meta-info"
            >
              <h1>
                System maintenance
              </h1>
            </div>
          </div>
        </div>
        <div
          class="action-menu-container"
        />
      </div>
    </div>
    <div
      class="tabs-container"
    >
      <div
        class="tabs "
      >
        <div
          class=" tab validations active busy"
        >
          <button
            class="tab-label"
          >
            Actions
          </button>
        </div>
        <div
          class=" tab cron busy"
        >
          <button
            class="tab-label"
          >
            Cron
          </button>
        </div>
        <div
          class=" tab seed busy"
        >
          <button
            class="tab-label"
          >
            Seed
          </button>
        </div>
        <div
          class=" tab database busy"
        >
          <button
            class="tab-label"
          >
            DB
          </button>
        </div>
        <div
          class=" tab activity busy"
        >
          <button
            class="tab-label"
          >
            Activity
          </button>
        </div>
        <div
          class=" tab plsc busy"
        >
          <button
            class="tab-label"
          >
            LDAP
          </button>
        </div>
        <div
          class=" tab composition busy"
        >
          <button
            class="tab-label"
          >
            Config
          </button>
        </div>
        <div
          class=" tab suspended-users busy"
        >
          <button
            class="tab-label"
          >
            Users
          </button>
        </div>
        <div
          class=" tab userlogins busy"
        >
          <button
            class="tab-label"
          >
            Logins
          </button>
        </div>
        <div
          class=" tab scim busy"
        >
          <button
            class="tab-label"
          >
            SCIM
          </button>
        </div>
        <div
          class=" tab stats busy"
        >
          <button
            class="tab-label"
          >
            Stats
          </button>
        </div>
        <div
          class=" tab pam busy"
        >
          <button
            class="tab-label"
          >
            PAM
          </button>
        </div>
        <div
          class=" tab proxy busy"
        >
          <button
            class="tab-label"
          >
            Proxy
          </button>
        </div>
        <div
          class=" tab sync busy"
        >
          <button
            class="tab-label"
          >
            Sync
          </button>
        </div>
      </div>
    </div>
    <div
      class="mod-system"
    >
      <section
        class="info-block-container"
      >
        <div
          class="mod-entities "
        >
          <section
            class="entities-search"
          >
            <h2>
              Open organisation invitations (0)
            </h2>
            <div
              class="search-filters orgInvitations-search-filters"
            />
            <div
              class="search standalone"
            >
              <div
                class="sds--text-field sds--text-field--has-icon"
              >
                <div
                  class="sds--text-field--shape"
                >
                  <div
                    class="sds--text-field--input-and-icon"
                  >
                    <input
                      class="sds--text-field--input"
                      placeholder="Search invitations..."
                      type="search"
                      value=""
                    />
                    <span
                      class="sds--text-field--icon"
                    >
                      <svg>
                        search.svg
                      </svg>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <section
            class="entities-list"
          />
          <div />
        </div>
      </section>
      <section
        class="info-block-container"
      >
        <div
          class="mod-entities "
        >
          <section
            class="entities-search"
          >
            <h2>
              Organisations without administrators (0)
            </h2>
            <div
              class="search-filters organisationsWithoutAdmin-search-filters"
            />
            <div
              class="search standalone"
            >
              <div
                class="sds--text-field sds--text-field--has-icon"
              >
                <div
                  class="sds--text-field--shape"
                >
                  <div
                    class="sds--text-field--input-and-icon"
                  >
                    <input
                      class="sds--text-field--input"
                      placeholder="Search organisations..."
                      type="search"
                      value=""
                    />
                    <span
                      class="sds--text-field--icon"
                    >
                      <svg>
                        search.svg
                      </svg>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <section
            class="entities-list"
          />
          <div />
        </div>
      </section>
      <section
        class="info-block-container"
      >
        <div
          class="mod-entities "
        >
          <section
            class="entities-search"
          >
            <h2>
              Applications without administrators (0)
            </h2>
            <div
              class="search-filters servicesWithoutAdmin-search-filters"
            />
            <div
              class="search standalone"
            >
              <div
                class="sds--text-field sds--text-field--has-icon"
              >
                <div
                  class="sds--text-field--shape"
                >
                  <div
                    class="sds--text-field--input-and-icon"
                  >
                    <input
                      class="sds--text-field--input"
                      placeholder="Search applications..."
                      type="search"
                      value=""
                    />
                    <span
                      class="sds--text-field--icon"
                    >
                      <svg>
                        search.svg
                      </svg>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </section>
          <section
            class="entities-list"
          />
          <div />
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
